components/script/dom/servoparser/async_html.rs

```diff
@@ -0,0 +1,561 @@
+        tree_builder.sink.trace(trc);
+    }
+}
+
+#[derive(JSTraceable, Clone, HeapSizeOf)]
+pub struct ParseNode {
+    id: usize,
+    qual_name: Option<QualName>,
+}
+
+enum ParserNodeOrText {
+    Node(usize),
+    Text(String),
+}
+
+#[derive(JSTraceable, HeapSizeOf)]
+struct ParseNodeData {
+    parent: Option<usize>,
+    target: Option<String>,
+    data: Option<String>,
+    contents: Option<ParseNode>,
+    is_integration_point: bool,
+}
+
+impl Default for ParseNodeData {
+    fn default() -> ParseNodeData {
+        ParseNodeData {
+            parent: None,
+            target: None,
+            data: None,
+            contents: None,
+            is_integration_point: false,
+        }
+    }
+}
+
+enum ParseOperation {
+    GetTemplateContents(usize, usize),
+    CreateElement(usize, QualName, Vec<Attribute>),
+    CreateComment(String, usize),
+    Insert(usize, Option<usize>, ParserNodeOrText),
+    AppendDoctypeToDocument(String, String, String),
+    AddAttrsIfMissing(usize, Vec<Attribute>),
+    RemoveFromParent(usize),
+    MarkScriptAlreadyStarted(usize),
+    ReparentChildren(usize, usize),
+    AssociateWithForm(usize, usize),
+    CreatePI(usize),
+    Pop(usize),
+}
+
+#[derive(JSTraceable, HeapSizeOf)]
+#[must_root]
+pub struct Sink {
+    base_url: ServoUrl,
+    document: JS<Document>,
+    current_line: u64,
+    script: MutNullableJS<HTMLScriptElement>,
+    parse_node_data: DOMRefCell<HashMap<usize, ParseNodeData>>,
+    next_parse_node_id: Cell<usize>,
+    nodes: DOMRefCell<HashMap<usize, JS<Node>>>,
+    document_node: ParseNode,
+}
+
+impl Sink {
+    fn new(base_url: ServoUrl, document: &Document) -> Sink {
+        let sink = Sink {
+            base_url: base_url,
+            document: JS::from_ref(document),
+            current_line: 1,
+            script: Default::default(),
+            parse_node_data: DOMRefCell::new(HashMap::new()),
+            next_parse_node_id: Cell::new(1),
+            nodes: DOMRefCell::new(HashMap::new()),
+            document_node: ParseNode {
+                id: 0,
+                qual_name: None,
+            }
+        };
+        let data: ParseNodeData = Default::default();
+        assert!(sink.parse_node_data.borrow_mut().insert(0, data).is_none());
+        sink.nodes.borrow_mut().insert(0, JS::from_ref(document.upcast()));
+        sink
+    }
+
+    fn new_parse_node(&self) -> ParseNode {
+        let id = self.next_parse_node_id.get();
+        let data: ParseNodeData = Default::default();
+        assert!(self.parse_node_data.borrow_mut().insert(id, data).is_none());
+        self.next_parse_node_id.set(id + 1);
+        ParseNode {
+            id: id,
+            qual_name: None,
+        }
+    }
+
+    fn enqueue(&self, op: ParseOperation) {
+        self.process_operation(op);
+    }
+
+    fn process_operation(&self, op: ParseOperation) {
+        let document = Root::from_ref(&**self.nodes.borrow().get(&0).expect("document node should exist!"));
```

Comment:
I wonder if the data for the document node shouldn't live outside the `nodes` map.